src/sysrepo.c

```diff
@@ -3561,40 +3561,34 @@ sr_delete_item(sr_session_ctx_t *session, const char *path, const sr_edit_option
     }
 
     if (session->ds == SR_DS_OPERATIONAL) {
-        if ((err_info = sr_lys_find_path(session->conn->ly_ctx, path, NULL, &snode))) {
-            /* invalid path */
+        set = NULL;
+        /* check that the xpath is valid */
+        if ((err_info = sr_lys_find_xpath(session->conn->ly_ctx, path, LYS_FIND_NO_MATCH_ERROR, NULL, &set))) {
             goto cleanup;
         }
-        if ((snode->nodetype & (LYS_LIST | LYS_LEAFLIST)) && (path[strlen(path) - 1] != ']')) {
-            /* purge all the (leaf-)list instances */
-            set = NULL;
-            if (session->dt[session->ds].edit->tree &&
-                    (err_info = sr_lyd_find_xpath(session->dt[session->ds].edit->tree, path, &set))) {
-                goto cleanup;
-            }
-            if (set && set->count) {
-                for (i = 0; i < set->count; ++i) {
-                    sr_lyd_free_tree_safe(set->dnodes[i], &session->dt[session->ds].edit->tree);
-                }
-            } else if (opts & SR_EDIT_STRICT) {
-                /* not found */
-                sr_errinfo_new(&err_info, SR_ERR_NOT_FOUND, "No nodes \"%s\" found in session push oper data.", path);
-            }
-            ly_set_free(set, NULL);
-        } else {
-            /* just delete the selected node */
-            node = NULL;
-            if (session->dt[session->ds].edit->tree &&
-                    (err_info = sr_lyd_find_path(session->dt[session->ds].edit->tree, path, 0, &node))) {
-                goto cleanup;
-            }
-            if (node) {
-                sr_lyd_free_tree_safe(node, &session->dt[session->ds].edit->tree);
-            } else if (opts & SR_EDIT_STRICT) {
-                /* not found */
-                sr_errinfo_new(&err_info, SR_ERR_NOT_FOUND, "Node \"%s\" not found in session push oper data.", path);
+        if (!set || !set->count) {
+            /* due to a libyang bug, LYS_FIND_NO_MATCH_ERROR doesn't seem to work correctly */
```

Comment:
If there is a bug in *libyang*, it can be fixed. Please provide the use-case and I can fix it.

Reply:
I think the commit 39a3459903a6252ef541259b70681a3aac601c20 I just added should reproduce the `libyang` bug and cause only a warning to be printed, in spite of 
```c
3565            if ((err_info = sr_lys_find_xpath(session->conn->ly_ctx, path, LYS_FIND_NO_MATCH_ERROR, NULL, &set))) {
(gdb)
767.909305 [3724081][140737347787520][WRN]: Schema node "list-enha" for parent "/mod:container" not found; in expr "/mod:container/list-enha" with context node "/".
3569            ly_set_free(set, NULL);
(gdb)
3570            set = NULL;

```

Reply:
Right, in should be fixed in the latest *libyang*.